File: implementations/v3-bun-htmx/backend/src/services/ReservationService.ts
import sql from "../database";
import type { Reservation } from "../entities/Reservation";
import { WSService } from "./WSService";

export class ReservationService {
    static async getActive(startStr: string, endStr: string): Promise<Reservation[]> {
        return await sql<Reservation[]>`
            SELECT * FROM reservations 
            WHERE status = 'active' 
            AND (start_time < ${endStr} AND end_time > ${startStr})
        `;
    }

    static async create(userId: number, spotId: number, start_time: string, end_time: string): Promise<void> {

        await sql.begin(async sql => {
            // 1. Lock
            const [spot] = await sql`SELECT id FROM spots WHERE id = ${spotId} FOR UPDATE`;
            if (!spot) throw new Error("Spot not found");

            // 2. Overlap
            const [count] = await sql`
                SELECT count(*) as count FROM reservations 
                WHERE spot_id = ${spotId} 
                AND status = 'active'
                AND (start_time < ${end_time} AND end_time > ${start_time})
            `;

            if (count.count > 0) throw new Error("Conflict");

            // 3. Insert
            await sql`
                INSERT INTO reservations (spot_id, user_id, start_time, end_time, status)
                VALUES (${spotId}, ${userId}, ${start_time}, ${end_time}, 'active')
            `;
        });

        WSService.broadcastUpdate(spotId);
    }

    static async release(id: number, userId: number): Promise<number> {
        let spotId = 0;
        await sql.begin(async sql => {
            const [res] = await sql`SELECT * FROM reservations WHERE id = ${id}`;
            if (!res) throw new Error("Reservation not found");
            if (res.user_id != userId) throw new Error("Not authorized");

            spotId = res.spot_id;
            await sql`DELETE FROM reservations WHERE id = ${id}`;
        });

        if (spotId) WSService.broadcastUpdate(spotId);
        return spotId;
    }

    static async getStats(): Promise<any[]> {
        return await sql`
            SELECT EXTRACT(HOUR FROM start_time) as hour, COUNT(*) as count 
            FROM reservations 
            GROUP BY hour 
            ORDER BY count DESC
        `;
    }
}

File: implementations/v2-python-vue/backend/services/reservation_service.py
import datetime
from fastapi import HTTPException
from entities.spot import Spot
from entities.reservation import Reservation
from services.ws_service import WebSocketService

class ReservationService:
    def __init__(self, conn):
        self.conn = conn
        self.spot_entity = Spot(conn)
        self.res_entity = Reservation(conn)

    def get_spots_with_reservations(self, date_str: str = None):
        if not date_str:
            date_str = datetime.date.today().isoformat()
        
        start_day = f"{date_str} 00:00:00"
        end_day = f"{date_str} 23:59:59"

        spots = self.spot_entity.get_all()
        reservations = self.res_entity.get_active_in_range(start_day, end_day)
        
        # Merge
        for spot in spots:
            spot['reservations'] = [dict(r) for r in reservations if r['spot_id'] == spot['id']] 

        return spots

    def create_reservation(self, user_id: int, spot_id: int, start_time: str, end_time: str):
        # Validate Time
        start_dt = datetime.datetime.fromisoformat(start_time)
        if start_dt < datetime.datetime.now() - datetime.timedelta(minutes=5):
            raise HTTPException(400, "Cannot book in the past")

        with self.conn.begin(): # Transaction
            # 1. Lock Spot
            spot = self.res_entity.lock_spot(spot_id)
            if not spot:
                raise HTTPException(404, "Spot not found")

            # 2. Check Overlap
            overlap = self.res_entity.count_overlaps(spot_id, start_time, end_time)
            if overlap > 0:
                raise HTTPException(409, "Spot already reserved")

            # 3. Insert
            self.res_entity.create(spot_id, user_id, start_time, end_time)
        
        WebSocketService.broadcast_update(spot_id)

    def complete_reservation(self, reservation_id: int, user_id: int):
        with self.conn.begin():
            res = self.res_entity.find_by_id(reservation_id)
            
            if not res:
                raise HTTPException(404, "Not found")
            
            if res['user_id'] != user_id:
                raise HTTPException(403, "Forbidden")
                
            self.res_entity.complete(reservation_id)
            
        WebSocketService.broadcast_update(res['spot_id'])

    def get_stats(self):
        return self.res_entity.get_hour_stats()
